[JIT] linear scan memory planning strategy

makslevental opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63873
* #63872
* #63539
* #63538
* __->__ #63537
* #63835
* #62174
* #62995

Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Commit:
[JIT] linear scan memory planning strategy

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,138 @@
+    const Region* reg = nullptr;
+    coalesce_avail(avail_regions);
+    for (auto& avail_reg : avail_regions) {
+      if (avail_reg.size >= aligned_curr_size) {
```
We might consider some other strategy here than taking first available region to avoid fragmentation. E.g. take the smallest available region that works. maybe at least add a todo to consider other strategies ?

eellison commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,138 @@
+            {reg->offset + aligned_curr_size, reg->size - aligned_curr_size});
+      }
+    } else {
+      // if possible spill smallest farthest out alloc
```
does spill in linear scan / register allocation means mean to move the value in register into memory. are we are doing something slightly different here ?

eellison commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,138 @@
+      if (!active.empty()) {
+        for (auto lv = active.end(); *lv != curr_range; lv--) {
+          auto alloced_reg = currently_alloced_regions[*lv];
+          if (alloced_reg.size >= aligned_curr_size) {
```
what is the point of swapping here ? the other region could have the same exact size as currently. also they might have different lifetimes.

eellison commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,138 @@
+            reg->offset, aligned_curr_size};
+        // split region (potentially)
+        if (reg->size - aligned_curr_size > 0) {
+          avail_regions.insert(
```
didnt we already do this above ? 

```
// split region (potentially)
```

eellison commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,138 @@
+      }
+
+      // swap i.e. put new alloc in old spot and malloc old alloc
+      if (reg != nullptr) {
```
Not sure I follow the logic here exactly... I dont know if we're doing enough checks on lifetime of swap_lvr.

eellison commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,138 @@
+  }
+
+  // last interval doesn't come back around to the top of the loop
+  expire_old_intervals(sorted_start_live_ranges.back());
```
yea sorry, why does we need to call this in addition to above ?

eellison reviewed (commented):
Nice!! Looking great. I kind of brain-dumped comments so some of them may not be accurate lol. 

My main question here is wrt to the spill intervals logic - in the original paper, that is meant as a way of dealing with a limited number of registers. We do not have the same constraint. Is it's inclusion here an optimization ? I don't fully follow.

eellison commented on tools/build_variables.bzl:
```diff
@@ -215,6 +215,7 @@ core_sources_full_mobile = [
     "torch/csrc/jit/passes/lower_grad_of.cpp",
     "torch/csrc/jit/passes/lower_tuples.cpp",
     "torch/csrc/jit/passes/memory_planning.cpp",
+    "torch/csrc/jit/passes/memory_planning/linear_scan.cpp",
```
More general case of pointwise operators

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]
